File: src/main/erlang/g0001_0100/s0034_find_first_and_last_position_of_element_in_sorted_array/Solution.erl
% #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Binary_Search
% #Algorithm_II_Day_1_Binary_Search #Binary_Search_I_Day_5 #Top_Interview_150_Binary_Search
% #Big_O_Time_O(log_n)_Space_O(1) #2025_01_12_Time_1_ms_(100.00%)_Space_69.07_MB_(100.00%)

-spec search_range(Nums :: [integer()], Target :: integer()) -> [integer()].
search_range(Nums, Target) ->
    [helper(Nums, Target, false), helper(Nums, Target, true)].

-spec helper(Nums :: [integer()], Target :: integer(), Equals :: boolean()) -> integer().
helper(Nums, Target, Equals) ->
    helper(Nums, Target, Equals, 0, length(Nums) - 1, -1).

-spec helper(Nums :: [integer()], Target :: integer(), Equals :: boolean(), L :: integer(), R :: integer(), Result :: integer()) -> integer().
helper(_, _, _, L, R, Result) when L > R ->
    Result;
helper(Nums, Target, Equals, L, R, Result) ->
    Mid = L + (R - L) div 2,
    MidValue = lists:nth(Mid + 1, Nums), % Erlang lists are 1-based
    NewResult = 
        case MidValue == Target of
            true -> Mid;
            false -> Result
        end,
    case MidValue < Target orelse (MidValue == Target andalso Equals) of
        true -> helper(Nums, Target, Equals, Mid + 1, R, NewResult);
        false -> helper(Nums, Target, Equals, L, Mid - 1, NewResult)
    end.

File: src/main/ruby/g0001_0100/s0034_find_first_and_last_position_of_element_in_sorted_array/solution.rb
# #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Binary_Search
# #Algorithm_II_Day_1_Binary_Search #Binary_Search_I_Day_5 #Top_Interview_150_Binary_Search
# #Big_O_Time_O(log_n)_Space_O(1) #2023_11_19_Time_60_ms_(77.52%)_Space_211.7_MB_(22.94%)

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def search_range(nums, target)
  ans = [helper(nums, target, false), helper(nums, target, true)]
  ans
end

private

def helper(nums, target, equals)
  l = 0
  r = nums.length - 1
  result = -1

  while l <= r
    mid = l + (r - l) / 2
    result = mid if nums[mid] == target
    if nums[mid] < target || (nums[mid] == target && equals)
      l = mid + 1
    else
      r = mid - 1
    end
  end

  result
end
